plugins/snowflake/dbt/include/snowflake/macros/adapters.sql

```diff
@@ -189,3 +189,14 @@
     {% endif %}
   {% endif %}
 {% endmacro %} 
+
+{% macro snowflake_dml_explicit_transaction(dml) %}
+  
+  {% set dml_transaction %}
+    begin;
+    {{ dml }} {{ ";" if not (dml | trim).endswith(";") }}
```

Comment:
This is really clever... worried that it might be too clever! If we wanted to be extra safe, we could issue three distinct queries for `begin`, the query, then `commit`. Would that defeat the purpose of these changes?

If we know that the supplied dml is always going to come from dbt internally, then this can be ok, but my preference would actually be either to:
1. Specify the semicolon in the caller OR
2. Always inject the semicolon here

you buy it?

Reply:
I thought about the latter option, as a new argument for the `statement` macro: perhaps `explicit_transaction=False`. It's conceptually related to the existing argument `auto_begin=True` (which won't take any effect on Snowflake, given the changes in this PR).

Since a lot of this DML is not called directly, however, but rather passed into the materialization's `main` statement—which could be DML, could be DDL—this would require a bit more plumbing work.

So I'm happy with mandating that SQL passed into this macro shouldn't end in a semicolon, and have the macro always stick it on the end.